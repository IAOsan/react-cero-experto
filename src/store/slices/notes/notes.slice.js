import { createSlice } from '@reduxjs/toolkit';
import { toast } from 'react-toastify';
import * as notesService from '../../../services/notes.service';
import * as imagesService from '../../../services/images.service';
import { convertFileToBase64 } from '../../../constants/utils';

const initialState = {
	list: [],
	active: null,
};

const slice = createSlice({
	name: '[notes]',
	initialState,
	reducers: {
		note_added(state, { payload }) {
			state.list.push(payload);
		},
		note_updated(state, { payload }) {
			const { id } = state.active;
			state.list = state.list.map((n) =>
				n.id === id ? { ...n, ...payload } : n
			);
		},
		note_changed(state, { payload }) {
			state.active = {
				...state.active,
				...payload,
			};
		},
		note_removed(state, { payload }) {
			state.list = state.list.filter((n) => n.id !== payload);
			state.active = null;
		},
		note_selected(state, { payload }) {
			const activeNote = state.list.find((n) => n.id === payload);
			state.active = activeNote;
		},
		notes_loaded(state, { payload }) {
			state.list = payload;
		},
		notes_cleared(state) {
			Object.keys(initialState).forEach((k) => {
				state[k] = initialState[k];
			});
		},
	},
});

export default slice.reducer;

const {
	note_added,
	note_changed,
	note_updated,
	note_selected,
	notes_loaded,
	note_removed,
	notes_cleared,
} = slice.actions;

export const addNote = async (dispatch, getState) => {
	try {
		const { uid } = getState().auth.user;
		const note = await notesService.addNewNote(uid);
		dispatch(note_added(note));
		dispatch(note_selected(note.id));
	} catch (error) {
		toast.error(error);
	}
};

export const changeNote = note_changed;

export const updateNote = (note) => async (dispatch, getState) => {
	try {
		const { uid } = getState().auth.user;

		if (note.file) {
			const meta = await imagesService.uploadImage(note.file);
			Object.assign(note, { imageUrl: meta.secure_url });
			delete note.file;
		}

		await notesService.updateNote(uid, note);
		dispatch(note_updated(note));
		toast.success('👏 Note saved!');
	} catch (error) {
		toast.error(error);
	}
};

export const selectNote = note_selected;

export const loadNotes = (uid) => async (dispatch) => {
	try {
		const notes = await notesService.getNotes(uid);
		dispatch(notes_loaded(notes));
	} catch (error) {
		toast.error(error);
	}
};

export const addImageToNote = (file) => async (dispatch) => {
	const base64 = await toast.promise(convertFileToBase64(file), {
		pending: 'Adding image, please wait...',
		success: 'Image added 👌',
		error: 'Sorry, the image could not be added 🤯, try uploading another one or try again later',
	});
	dispatch(note_changed({ imageUrl: base64 }));
};

export const removeNote = (id) => async (dispatch, getState) => {
	try {
		const { uid } = getState().auth.user;
		await notesService.removeNote(uid, id);
		dispatch(note_removed(id));
		toast.success('👍 Note removed!');
	} catch (error) {
		toast.error(error);
	}
};

export const clearNotes = notes_cleared();
