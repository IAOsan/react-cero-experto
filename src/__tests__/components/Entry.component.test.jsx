import '@testing-library/jest-dom';
import { vi } from 'vitest';
import { renderWithProviders, setupUser } from '../test-utils';
import Entry, { generateEntryDate } from '../../components/Entry.component';
import { selectNote } from '../../store/slices/notes/notes.slice';

vi.mock('../../store/slices/notes/notes.slice', async () => {
	const org = await vi.importActual('../../store/slices/notes/notes.slice');
	return {
		...org,
		selectNote: vi.fn(() => vi.fn()),
	};
});

const entry = {
	id: 1,
	title: 'title',
	body: 'body',
	createdAt: new Date().toISOString(),
};
const user = setupUser();

describe('generateEntryDate()', () => {
	it('should return an object with day and date formated', () => {
		const date = new Date().toISOString();
		const expectedResult = {
			date: new Date(date).getDate(),
			day: new Date(date).toLocaleDateString('en-US', {
				weekday: 'long',
			}),
		};

		const result = generateEntryDate(date);

		expect(result).toEqual(expectedResult);
	});
});

describe('<Entry />', () => {
	beforeEach(() => {
		vi.clearAllMocks();
	});

	it('should renders properly', () => {
		const { date, day } = generateEntryDate(entry.createdAt);

		const wrapper = renderEntry({ ...entry });

		expect(wrapper.container.firstElementChild).toMatchSnapshot();
		expect(wrapper.getByText(entry.title)).toBeInTheDocument();
		expect(wrapper.getByText(entry.body)).toBeInTheDocument();
		expect(wrapper.getByText(day)).toBeInTheDocument();
		expect(wrapper.getByText(date)).toBeInTheDocument();
	});

	it('should renders image if exists', () => {
		Object.assign(entry, { imageUrl: 'https://via.placeholder.com/150C/' });

		renderEntry({ ...entry });

		const imageElement = document.querySelector('.entry__img-box');

		expect(imageElement).toBeInTheDocument();
		expect(imageElement).toHaveStyle(`--bg: url(${entry.imageUrl})`);
	});

	it('should fire selectNote action when clicks on it', async () => {
		const wrapper = renderEntry({ ...entry });

		await user.click(wrapper.getByRole('button'));

		expect(selectNote).toHaveBeenCalledWith(entry.id);
	});
});

function renderEntry(props = {}) {
	return renderWithProviders(<Entry {...props} />);
}
