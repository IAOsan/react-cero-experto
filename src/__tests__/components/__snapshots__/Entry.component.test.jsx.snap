// Vitest Snapshot v1

exports[`<Entry /> > should renders properly 1`] = `
<li
  class="anima-fade-in-bottom"
>
  <button
    class="btn w-100 flex entry"
    type="button"
  >
    <div
      class="entry__body"
    >
      <h3
        class="font-body mb-8"
      >
        title
      </h3>
      <p>
        body
      </p>
    </div>
    <div
      class="flex flex-ai-c flex-jc-c text-center entry__date"
    >
      <p>
        <span>
          Tuesday
        </span>
        <br />
        <strong>
          30
        </strong>
      </p>
    </div>
  </button>
</li>
`;
